### PR TITLE
docs: add parsing request

### PR DESCRIPTION
**Pull request by jannyHou**
Connect to https://github.com/strongloop/loopback-next/pull/1489

The story only contains validation doc, while I realized the doc for action `parsing request` is missing, so I added it and also the coercion doc.

## Checklist

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: docs/site/Parsing-requests.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+- Coerces parameters from string to its corresponding JavaScript run-time type.
+- Performs validation on the parameters and body data.
+
+### Parsing Raw Data
+
+The code below defines a typical endpoint by decorating a controller method with
+rest decorators.
+
+```ts
+class TodoController() {
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: nitpick: `this.todoRepo` is not specified in this TS class

File: docs/site/Parsing-requests.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    @param.path.number('id') id: number,
+    @requestBody() todo: Todo,
+  ): Promise<boolean> {
+    return await this.todoRepo.replaceById(id, todo);
+  }
+}
+```
+
+An operation specification will be generated in-memory to describe it, and raw
+datas are parsed from request according to the specification. For example the
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: data does not have a plural form

Comment (shimks) [reply docs/site/Parsing-requests.md]: Did you mean `For example, if the first parameter ...`?

File: docs/site/Parsing-requests.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+_See [controller](Controller.md) for more details of defining an endpoint._
+
+_See
+[OpenAPI operation object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#operationObject)
+to know more about its structure._
+
+### Coercion
+
+The parameters parsed from path, header, query of a http request are always in
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `path, header and query` (American way of using and)

Comment (shimks) [reply docs/site/Parsing-requests.md]: It also might be good to provide context that it's **node.js**'s request parameters being in string format

File: docs/site/Parsing-requests.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+the string format. Therefore when invoking a controller function, a parameter
+need to be converted to its corresponding JavaScript runtime type, which is
+inferred from its parameter specification.
+
+Give the example operation `replaceTodo` in section
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `Take` instead of `Give` might be a better choice of word

File: docs/site/Parsing-requests.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+[parsing raw data](#parsing-raw-data), which takes in a number `id` as the first
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: no comma here

File: docs/site/Parsing-requests.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+input, coercion means users don't need to do a type conversion by themselves in
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: period instead of comma here.
I also think the latter sentence can be better phrased. Maybe something like `Without coercion, `id` would have to be manually cast into the number type before it can be used as seen below:`

File: docs/site/Parsing-requests.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+keywords: LoopBack 4.0, LoopBack 4
+tags:
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Parsing-requests.html
+summary:
+---
+
+## Parsing Requests
+
+This is an action in the default HTTP sequence, it parses arguments from an
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: Can you specify what the key that the action is bound to is (PARSE_PARAMS)? It might be good to see it in code as well

File: docs/site/Parsing-requests.md (hunk 1 above)
Comment (shimks) [reply docs/site/Parsing-requests.md]: No parenthesis

File: docs/site/Parsing-requests.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+### Validation
+
+Validations are applied on the parameters and the request body data. They also
+use OpenAPI specification as the reference to infer the validation rules.
+
+#### Parameters
+
+We have the data type safety check for the parameters parsed from header, path,
+query. For example, if a parameter should be an integer, then a number with
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `header, path and query`

File: docs/site/Parsing-requests.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+decimal like "1.23" will be rejected.
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: would

File: docs/site/Parsing-requests.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+You can specify a parameter's type by calling shortcut decorators of `@param`
+like `@param.query.integer()`. A list of available shortcuts can be found in the
+[api documents](https://apidocs.strongloop.com/@loopback%2fdocs/openapi-v3.html#param).
+And please check [parameter decorators](Decorators.md#parameter-decorator) for
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `Check out the section on parameter decorators(link) for instructions on how to decorate the ...`

File: docs/site/Parsing-requests.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  values: `TRUE`, `1`, `FALSE` or `0`.
+
+#### Request Body
+
+The data from request body is validated against its OpenAPI schema
+specification. We use module [AJV](https://github.com/epoberezkin/ajv) to
+perform the validation, which validates data with a JSON schema generated from
+the OpenAPI schema specification.
+
+Take operation `replaceTodo` as an instance again:
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `Take again the operation replaceTodo for instance`

File: docs/site/Parsing-requests.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+```ts
+import {Todo} from './models/todo-model.ts';
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: nitpick: `./models`

File: docs/site/Parsing-requests.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    return await this.todoRepo.replaceById(id, todo);
+  }
+...
+```
+
+The request body specification is defined by applying `@requestBody()` to
+argument `todo`, and the schema specification inside it is inferred from its
+type `Todo`. The type is exported from a `Todo` model.
+
+_See [model](Model.md) to know more details about how to decorate a model class_
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: We should wrap this around `note.html`

File: docs/site/Parsing-requests.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+When the `PUT` method `/todo/{id}` get called, the `todo` instance from request
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `gets`

Comment (shimks) [reply docs/site/Parsing-requests.md]: `the request body`

File: docs/site/Parsing-requests.md [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+body will be validated with a well defined specification.
+
+Section [RequestBody Decorator](Decorators.md#requestbody-decorator) has a very
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: Section on requestBody Decorator

File: docs/site/Parsing-requests.md [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+detailed explanation of the various ways to provide a request body
+specification. It would be a redundant to repeat it here. So make sure you read
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: I don't think we need the last two sentences

File: docs/site/Parsing-requests.md [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+- If a model property's type refers to another model, make sure it is also
+  decorated with `@model` decorator.
+
+- By the API first approach, you can also provide the request body specification
+  in decorators like `route()` and [`api()`](Decorators.md#api-decorator), this
+  requires you to provide a completed request body specification.
+
+#### Localizing errors
+
+A body data may contain multiple invalid things, like missing required field,
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `A body data could break multiple validation rules, ...` `things` is I think too informal here.

File: docs/site/Sequence.md [hunk 18]
```diff
@@ -229,7 +229,7 @@ How to get query string param values.
 
 ### Parsing Requests
```
Comment (shimks) [inline docs/site/Sequence.md]: If we're going to have a subpage for this section, we should get rid of its mention entirely here.
Maybe we can also consider adding a link

File: docs/site/Parsing-requests.md (hunk 2 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: I meant the example code above, rephrased it to:
> In the example
above, the first parameter is from source `path`, then its value will be parsed
from a request's path.

File: docs/site/Parsing-requests.md (hunk 5 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: combined with the comment above, do you mean rephrase it to:

> Take the example operation `replaceTodo` in section
[parsing raw data](#parsing-raw-data) which takes in a number `id` as the first
input, coercion means users don't need to do a type conversion by themselves in
the controller function like below:

?

File: docs/site/Parsing-requests.md (hunk 7 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: I am not sure what's the point of introducing the binding key here...?

File: docs/site/Parsing-requests.md (hunk 12 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: good catch!

File: docs/site/Parsing-requests.md (hunk 16 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: hmm I do want to emphasize "make sure you read that section", I deleted the fist sentence.

File: docs/site/Sequence.md (hunk 18 above)
Comment (jannyHou) [reply docs/site/Sequence.md]: important catch! thx

File: docs/site/Parsing-requests.md [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+    @requestBody() todo: Todo,
+  ): Promise<boolean> {
+    return await this.todoRepo.replaceById(id, todo);
+  }
+}
+```
+
+An operation specification will be generated in-memory to describe it, and raw
+data are parsed from request according to the specification. In the example
+above, the first parameter is from source `path`, then its value will be parsed
```
Comment (shimks) [inline docs/site/Parsing-requests.md]: `path, so its value will be parsed from a request's path`

File: docs/site/Parsing-requests.md (hunk 5 above)
Comment (shimks) [reply docs/site/Parsing-requests.md]: Nevermind, please put the comma back in. 🙇

File: docs/site/Parsing-requests.md [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+  async replaceTodo(
+    @param.path.number('id') id: number,
+    @requestBody() todo: Todo,
+  ): Promise<boolean> {
+    return await this.todoRepo.replaceById(id, todo);
+  }
+}
+```
+
+An operation specification will be generated in-memory to describe it, and raw
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: nitpick: `An OpenAPI operation specification`

File: docs/site/Parsing-requests.md [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+data are parsed from request according to the specification. In the example
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: another nitpick: `and raw data are parsed from request` -> `and raw data is parsed from the request`

File: docs/site/Parsing-requests.md (hunk 7 above)
Comment (shimks) [reply docs/site/Parsing-requests.md]: The coercion/validation is done by `parser.ts`'s `parseOperationArgs`, which is bound by default to the key `RestBindings.SequenceActions.PARSE_PARAMS`. Without this action being triggered int he sequence, no validation/coercion would be done.

The section is also a child page of `Sequence.md`, so might as well

File: docs/site/Parsing-requests.md (hunk 16 above)
Comment (shimks) [reply docs/site/Parsing-requests.md]: How about this: `Validation of model objects is heavily dependent on its OpenAPI Schema defined in/by @requestBody decorator. Please refer to the documentation on [@requestBody decorator](link) to get a comprehensive idea of defining custom validation rules for your models.`

File: docs/site/Parsing-requests.md [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+
+#### Parameters
+
+We have the data type safety check for the parameters parsed from header, path,
+and query. For example, if a parameter should be an integer, then a number with
+decimal like "1.23" would be rejected.
+
+You can specify a parameter's type by calling shortcut decorators of `@param`
+like `@param.query.integer()`. A list of available shortcuts can be found in the
+[api documents](https://apidocs.strongloop.com/@loopback%2fdocs/openapi-v3.html#param).
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: nitpick: `api documents` -> `API Docs`

File: docs/site/Parsing-requests.md [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+Check out the section on
+[parameter decorators](Decorators.md#parameter-decorator) for instructions on
+how to decorate the controller parameter.
+
+Here are our default validation rules for each type:
+
+- number: validated by `isNaN(Number(data))`.
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: nitpick: I suggest to either remove `.` for each line of the list or add it to be consistent

File: docs/site/Parsing-requests.md [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+  [RFC3339](https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14).
+- date: should be a valid full-date defined in
+  [RFC3339](https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14).
+- boolean: after converted to all upper case, should be one of the following
+  values: `TRUE`, `1`, `FALSE` or `0`.
+
+#### Request Body
+
+The data from request body is validated against its OpenAPI schema
+specification. We use module [AJV](https://github.com/epoberezkin/ajv) to
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: nitpick: `module AJV` -> `AJV module`

File: docs/site/Parsing-requests.md [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+perform the validation, which validates data with a JSON schema generated from
+the OpenAPI schema specification.
+
+Take again the operation replaceTodo for instance:
+
+```ts
+import {Todo} from './models';
+
+...
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: minor nitpick: can we wrap this code block with the controller class declaration?

```ts
import {Todo} from './models';

class test {
  ...
  @put('/todos/{id}')
  async replaceTodo(
    @param.path.number('id') id: number,
    @requestBody() todo: Todo,
  ): Promise<boolean> {
    return await this.todoRepo.replaceById(id, todo);
  }
  ...
}

```

it just didn't render properly for me in VS code :P

File: docs/site/Parsing-requests.md [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+The request body specification is defined by applying `@requestBody()` to
+argument `todo`, and the schema specification inside it is inferred from its
+type `Todo`. The type is exported from a `Todo` model.
+
+{% include note.html title="Model documentation" content="See [model](Model.md) to know more details about how to decorate a model class" %}
+
+When the `PUT` method `/todo/{id}` gets called, the `todo` instance from the
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: nitpick: `'PUT' method on path '/todo/{id}'`

File: docs/site/Parsing-requests.md [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+very detailed explanation of the various ways to provide a request body
+specification. Make sure you read that section to understand how to use
+`@requestBody()`.
+
+A few tips worth mentioning:
+
+- If a model property's type refers to another model, make sure it is also
+  decorated with `@model` decorator.
+
+- By the API first approach, you can also provide the request body specification
```
Comment (b-admike) [inline docs/site/Parsing-requests.md]: nitpick: `By the API first approach` -> `If you're using API first development approach`

File: docs/site/Parsing-requests.md (hunk 7 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: I see...my doc only covers the default behaviour of this action, and I think those actions would be bound automatically after extending `Application` with the rest mixin :)

I would like to add a section on `sequence` for replacing/customizing the actions, and list the binding key for each action under it, if such a section doesn't exist. Otherwise we need to repeat these things in every action: in which situations they come out of the box, in which situations 
it's on the users to bind them manually, and how to replace them with a custom action.

thought?

File: docs/site/Parsing-requests.md (hunk 16 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: cool 😎 !

File: docs/site/Parsing-requests.md (hunk 23 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: @bajtos hmm...I added it for each line....on which line you didn't find it?

File: docs/site/Parsing-requests.md (hunk 25 above)
Comment (jannyHou) [reply docs/site/Parsing-requests.md]: There are too many irrelevant code in the class definition which distract users when reading :( that's why I deleted them.

> it just didn't render properly for me in VS code

Mine looks good 🤔 :
screen shot 2018-07-13 at 3 11 18 pm
